ee/clickhouse/views/test/test_clickhouse_retention.py

```diff
@@ -40,45 +97,135 @@ def test_can_get_line_chart_and_fetch_people(self):
+            },
+        )
 
+    def test_can_specify_breakdown_event_property(self):
+        """
+        By default, we group users together by the first time they perform the
+        `target_event`. However, we should also be able to specify, e.g. the
+        users OS to be able to compare retention between the OSs.
+        """
+        organization = create_organization(name="test")
+        team = create_team(organization=organization)
+        user = create_user(email="test@posthog.com", password="1234", organization=organization)
 
-class RetentionRequest(TypedDict):
+        self.client.force_login(user)
+
+        update_or_create_person(distinct_ids=["person 1"], team_id=team.pk)
+        update_or_create_person(distinct_ids=["person 2"], team_id=team.pk)
+
+        setup_user_activity_by_day(
+            daily_activity={
+                "2020-01-01": {"person 1": [{"event": "target event", "properties": {"os": "Chrome"}}]},
+                "2020-01-02": {
+                    "person 1": [{"event": "target event"}],
+                    "person 2": [{"event": "target event", "properties": {"os": "Safari"}}],
+                },
+                # IMPORTANT: we include data past the end of the requested
+                # window, as we want to ensure that we pick up all retention
+                # periods for a user. e.g. for "person 2" we do not want to miss
+                # the count from 2020-01-03 e.g. the second period, otherwise we
+                # will skew results for users that didn't perform their target
+                # event right at the beginning of the requested range.
+                "2020-01-03": {"person 1": [{"event": "target event"}], "person 2": [{"event": "target event"}]},
+            },
+            team=team,
+        )
+
+        retention = get_retention_ok(
+            client=self.client,
+            team_id=team.pk,
+            request=RetentionRequest(
+                target_entity={"id": "target event", "type": "events"},
+                returning_entity={"id": "target event", "type": "events"},
+                date_from="2020-01-01",
+                total_intervals=2,
+                date_to="2020-01-02",
+                period="Day",
+                retention_type="retention_first_time",
+                breakdowns=[Breakdown(type="event", property="os")],
+                # NOTE: we need to specify breakdown_type as well, as the
+                # breakdown logic currently does not support multiple differing
+                # types
+                breakdown_type="event",
+            ),
+        )
+
+        retention_by_cohort_by_period = get_by_cohort_by_period_from_response(response=retention)
+
+        self.assertEqual(
+            retention_by_cohort_by_period,
+            {
+                "Chrome": {"1": 1, "2": 1},
+                "Safari": {"1": 1, "2": 1},  # IMPORTANT: the "2" value is from past the requested `date_to`
```

Comment:
already discussed this but adding as a note here for documentation. We'll need to consider how to update retention UI to handle breakdown values as it's adding an extra dimension to consider in the activity matrix

Reply:
Yup I'll follow up with the frontend.

This change actually just changes the dimension to breakdown value from cohort acquisition date, but with the change to also have the cohort breakdown it will certainly add complexity to the response and frontend.